feat: 앱 리뷰 팝업 창을 띄우도록 기능을 추가해요

Do-hyun-Kim opened the pull request:
## 🔵PR을 올리기 전 아래 사항을 확인해주세요.
- 구현한 로직과 기능이 올바르게 작동되는지 충분히 테스트해주세요.
- 코드의 성능이나 메모리 효율성이 적절하게 고려되었는지, 불필요한 코드가 없는지 검토해주세요.
- 이번 PR에서 구현된 주요 기능이나 해결된 문제에 대해 자세히 서술해주세요.
(위 내용은 지워주세요)



## 😽개요

* 포스트 업로드 이후 별점 팝업 띄우는 로직 추가해요
* 이전 `DailyCalendarViewController`에 가족이 모두 올린 날에는 fireLottieView가 발생했지만 사라져 있어서 추가했습니다.
* 일부 불필요한 코드 조금 제거했습니다.
* `FetchUserCreateAtInfoUseCase ` 추가하여 팝업 띄우는 로직을 추가했습니다.
* `FCMToken` Local Storage에 저장하는 로직 추가

## 🛠️FetchUserCreateAtInfoUseCase 내부 `사용자 가입 일자`, `마지막 리뷰 날짜`, `최신 버전`, `리뷰한 횟수` 기준으로 판단 여부 로직 추가
- 기존 사용자 같은 경우 `사용자 가입 일자`는 서버로부터 데이터를 받을 수 있지만, `마지막 리뷰 날짜` 같은 경우는 데이터가 없기 때문에 생성 일자 기준으로 넣어놨습니다.
- 리뷰한 횟수는 3개를 넘지 않도록 했습니다(정책 상 1년 동안 리뷰 팝업을 3번 노출하지 않아야 한다 해서).


### FetchUserCreateAtInfoUseCase 로직 MainViewReactor 추가
- `Survival` 게시글 업로드 이후 팝업창을 띄우도록 하려 하는데 혹시 이 부분이 맞을지 확인 부탁드릴게용
```swift
/// MainViewReactor(Action)
        case .checkMissionAlert(let isUnlocked, let isMeSurvivalUploadedToday):
            if !(isUnlocked && isMeSurvivalUploadedToday) {
                return .empty()
            }
            
            return checkMissionAlertShowUseCase.execute()
                .filter { !$0 }
                .withUnretained(self)
                .flatMap { owner, isAlreadyShown -> Observable<Mutation> in
                    return owner.fetchUserCreateAtUseCase.execute()
                        .flatMap { isCheck -> Observable<Mutation> in
                            owner.pushViewController(type: .missionUnlockedAlert)
                            return .just(.setRatingAlert(isCheck))
                        }
                }
```

## ✅테스트 케이스

* 팝업 창이 로직에 구성에 맞게 잘 뜨는지 확인해요(테플 업로드 이후 확인 가능) 디버그 모드는 실시간으로 뜬다고 하네요
* DailyCalendarViewController 가족 모두 업로드 시 Lottie가 잘 작동하는지 확인

---

#### 🙏🏻아래와 같이 PR을 리뷰해주세요.
- PR 내용이 부족하다면 보충 요청해주세요.
- 코드 스타일이 팀의 규칙에 맞게 작성되었는지, 일관성을 유지하고 있는지 확인해주세요.
- 코드에 대한 문서화나 주석이 필요한 부분에 적절하게 작성되어 있는지 확인해주세요.
- 구현된 로직이 효율적이고 올바르게 작성되었는지, 아키텍처를 잘 준수하고 있는지 검토해주세요.
- 네이밍, 포매팅, 주석 등 코드의 일관성이 유지되고 있는지 확인해주세요.

Do-hyun-Kim committed:
feat: MyDIContainer FetchUserCreateAtInfoUseCase 의존성 주입 코드 추가
- Tuist 음성 권한 팝업 문구 수정
- UserDefaultsKey 최신버전, 리뷰 카운트, 마지막 업데이트 날짜 저장 Key 등록
- MyRepository 최신버전, 리뷰 카운트, 마지막 업데이트 날짜 Update, Fetch 메서드 정의
- MyUserDefaults 내부 최신버전, 리뷰 카운트, 업데이트 날짜 저장 메서드 추가
- FetchUserCreateAtInfoUseCase 리뷰 팝업 정책에 따른 로직 추가

Do-hyun-Kim committed:
fix: MainNavigator Alert 사라지지 않는 이슈 수정
- MainViewController 초기 NoEmptyView 안뜨는 이슈 수정

Do-hyun-Kim committed:
feat: Extensions 폴더 내부 SKStoreReviewController+Ext 파일 및 리뷰 팝업 메서드 추가

Do-hyun-Kim committed:
feat: MainViewReactor 별점 팝업 관련 State, UseCase 추가
- MainViewController isShowRatingAlert State에따라 팝업 띄우는 로직 추가
- ProfileViewReactor 테스트 코드 제거
- UserDefaultsKey isLatestVersion Key 추가, latestVersion 제거
- MyRepository fetchIsLatestVersion 메서드 추가, fetchLatestVersion 메서드 제거
- MeAPIWorker latest 필드 값 저장 로직 추가
- fetchUserCreateAtInfoUseCase 로직 수정

akrudal commented on 14th-team5-iOS/Bibbi/Sources/Presentation/Home/ViewControllers/MainViewController.swift:
```diff
@@ -186,6 +187,13 @@ extension MainViewController {
+        reactor.pulse(\.$isShowRatingAlert)
+            .filter { $0 }
+            .bind { _ in
+                SKStoreReviewController.requestReviewInCurrentScene()
```
이거 있잖아요.. 이거 뭐 꼭 홈이여야 할까요? 이런 팝업 띄우고 하는게 꼭 홈만의 책임은 아니라 생각하는데
이거 최상단 네비게이션 바에서 띄우는건 어때요? 사실 이거말고도 그냥 토스트 메세지 생각해봐도 다음 뷰로 넘어가더라도 토스트메세지는 남아있어야하잖아여...? 그런거를 관리하기 위해 최상단을 하나 더 두는게 어떨지

akrudal replied:
이게 맞는 것 같은게 사진찍고 성공하고, 네비게이터에 전달해서 가족 생성일 체크하고 팝업 띄우는게 맞는것 같네욥..?

왜냐면 지금 미션체크하는거에 그걸 두신것 같은데 그거랑 생존신고 한번하고 띄우는거랑은 차이가 있지않나요

미션은 4명가족이면 두명까지 올려야 확인이되어가지고 그때 리뷰팝업이 뜨게되는거아닌가엽?

Do-hyun-Kim replied:
> 이거 있잖아요.. 이거 뭐 꼭 홈이여야 할까요? 이런 팝업 띄우고 하는게 꼭 홈만의 책임은 아니라 생각하는데 이거 최상단 네비게이션 바에서 띄우는건 어때요? 사실 이거말고도 그냥 토스트 메세지 생각해봐도 다음 뷰로 넘어가더라도 토스트메세지는 남아있어야하잖아여...? 그런거를 관리하기 위해 최상단을 하나 더 두는게 어떨지
- 최상단 내비게이션 바에 띄우자는 게 어떤 말씀이실 까요?? 최상위 뷰에서 띄우자는 것일까요??

Do-hyun-Kim replied:
> 이게 맞는 것 같은게 사진찍고 성공하고, 네비게이터에 전달해서 가족 생성일 체크하고 팝업 띄우는게 맞는것 같네욥..?
> 
> 왜냐면 지금 미션체크하는거에 그걸 두신것 같은데 그거랑 생존신고 한번하고 띄우는거랑은 차이가 있지않나요
> 
> 미션은 4명가족이면 두명까지 올려야 확인이되어가지고 그때 리뷰팝업이 뜨게되는거아닌가엽?

오옹 맞네용 (best practice!!) 그럼 카메라 사진 찍고 나서 UseCase로 가족 생성일 체크하고 팝업 띄우는 형식으로 하겠습니다.

akrudal replied:
엄... 최상단 네비게이션이 맞는 것 같네욥...?!

akrudal replied:
특정 뷰의 역할은 아닐 것 같아서요....!!! 어케 생각하시는지?!

Do-hyun-Kim replied:
생각 해보니깐 저 SKStoreReviewController는 최상단에서 띄우고 있어용 `UIApplication` 에 `WindowScene` 을 통해서 띄우도록 되어있어서

Do-hyun-Kim committed:
fix: CameraDisplayViewReactor Survival 피드 업로드 이후 RatingUseCase 로직 실행 구문 추가
- MainViewControllerWrapper isRatingAlertHidden Flag 값 추가
- CameraDisplayNavigator isRatingHidden Parameter 추가
- AccountResingViewReactor FCMToken 제거 API 추가 및 Keychain Storage FCM Token 제거 로직 추가
- MeAPIWorker FCMToken Local Storage에 저장하는 로직 추가

Do-hyun-Kim committed:
Merge branch 'develop' into feat/#753-add-store-review-alert

Do-hyun-Kim committed:
chore: github Action Push braches Event 실행 하도록 추가
- Xcode 16.2 환경으로 실행 하도록 수정

akrudal commented on 14th-team5-iOS/Bibbi/Sources/Application/DIContainer/MyDIContainer.swift:
```diff
@@ -61,6 +61,10 @@ final class MyDIContainer: BaseContainer {
     }
     
+    private func makeFetchUserCreateAtInfoUseCase() -> FetchUserCreateAtInfoUseCaseProtocol {
+        return FetchUserCreateAtInfoUseCase(myRepository: makeMyRepository())
```
zzzzz이거 제가 저번에 커멘트 깜빡햇는데.... createdAt 으로 해주실 수. ㅣㅆ나염..?😂

Do-hyun-Kim replied:
네넵 수정해서 올리겠습니당 :)

Do-hyun-Kim committed:
fix: FetchUserCreatedAtInfoUseCase로 네이밍 수정
- MyRepository 내부 fetchFamilyCreateAt 메서드 네이밍 CreatedAt 메서드로 네이밍 수정
- MyDIContainer FetchUserCreatedAtInfoUseCaseProtocol 및 makeFetchUserCreatedAtInfoUseCase로 네이밍 수정

akrudal reviewed (approved):
고생하셨슴다
